File: LC/python/Reverse Linked List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @param m, an integer
    # @param n, an integer
    # @return a ListNode
    def reverseBetween(self, head, m, n):
        if head == None or head.next == None:
            return head
        
        node = []
        
        start = tmp = head
        
        for i in range(m-1):
            start = start.next
            tmp = tmp.next
            
        for i in range(n-m+1):
            node.append(start.val)
            start = start.next
            
        for i in range(n-m+1):
            tmp.val = node[-1-i]
            tmp = tmp.next
        
        return head
        

File: LC/java/Reverse Linked List II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        if(head==null || m>=n){
            return head;
        }
        
        ListNode p, mp, a, b, temp;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        p = dummy;
        
        for(int i = 1; i < m; i++){
            if(p == null){
                return null;
            }
            p = p.next;
        }
        
        mp = a = p.next;
        b = a.next;
        for(int i = m; i < n; i++){
            if(b == null){
                return null;
            }
            temp = b.next;
            b.next = a;
            a = b;
            b = temp;
        }
        mp.next = b;
        p.next = a;
        
        return dummy.next;
        
    }         
}